source/common/upstream/upstream_impl.h

```diff
@@ -480,10 +480,15 @@ class PrioritySetImpl : public PrioritySet {
   void updateHosts(uint32_t priority, UpdateHostsParams&& update_hosts_params,
                    LocalityWeightsConstSharedPtr locality_weights, const HostVector& hosts_added,
                    const HostVector& hosts_removed,
-                   absl::optional<uint32_t> overprovisioning_factor = absl::nullopt) override;
+                   absl::optional<uint32_t> overprovisioning_factor = absl::nullopt,
+                   HostMapConstSharedPtr cross_priority_host_map = nullptr) override;
 
   void batchHostUpdate(BatchUpdateCb& callback) override;
 
+  const HostMapConstSharedPtr& crossPriorityHostMap() const override {
```

Comment:
Should this return by value? Seems a bit safer so nothing stores this by ref and tries to access it?

Reply:
I think if someone needs to store it, then a copy of the shared pointer should be created from this ref instead of storing the reference directly. 🤔 

According to my imagination, this method has two main usage scenarios:
- First, in the main thread, we use this method to get the latest host map. At this point, as the code shows, I will create a copy of the shared pointer.
- Second, in the worker threads, we will obtain a reference to the host map through this method, and then perform a host search, but will not save this reference. In this case, returning reference can reduce unnecessary atomic operations.

Reply:
So is this in preparation for another usage where we access it without taking a copy from the worker threads? All the usages in this PR seems to make a copy. In general forcing a copy via the API makes the API a bit safer since it means the caller can't mess up by trying to access the ref later on, so it makes it less likely for bugs to happen later

If we are trying to allow for ref access via the ptr I think we need to add some comments around the lifetime of the returned reference since it is different from the lifetime of the PSet

Another strategy would be to have a `std::shared_ptr<T>` and a `const T&` accessor, which lets consumers of the API choose between a call that has very clear ownership semantics and one that requires some special attention

Reply:
> So is this in preparation for another usage where we access it without taking a copy from the worker threads?

Yes, it is preparation for Sticky LB.

🤔 My be we can return shared pointer directly. As long as we don't call this method frequently, there won't be too many atomic operations.